packages/cli/src/cmds/validator/keymanager/impl.ts

```diff
@@ -142,30 +145,43 @@ export class KeymanagerApi implements Api {
           continue;
         }
 
-        // Attempt to decrypt before writing to disk
-        const secretKey = bls.SecretKey.fromBytes(await keystore.decrypt(password));
-
-        // Persist the key to disk for restarts, before adding to in-memory store
-        // If the keystore exist and has a lock it will throw
-        this.persistedKeysBackend.writeKeystore({
-          keystoreStr,
-          password,
-          // Lock immediately since it's gonna be used
-          lockBeforeWrite: true,
-          // Always write, even if it's already persisted for consistency.
-          // The in-memory validatorStore is the ground truth to decide duplicates
-          persistIfDuplicate: true,
-        });
-
-        // Add to in-memory store to start validating immediately
-        this.validator.validatorStore.addSigner({type: SignerType.Local, secretKey});
-
-        statuses[i] = {status: ImportStatus.imported};
+        decryptKeystores.queue(
+          {keystoreStr, password},
+          (secretKeyBytes: Uint8Array) => {
+            try {
+              const secretKey = bls.SecretKey.fromBytes(secretKeyBytes);
+
+              // Persist the key to disk for restarts, before adding to in-memory store
+              // If the keystore exist and has a lock it will throw
+              this.persistedKeysBackend.writeKeystore({
+                keystoreStr,
+                password,
+                // Lock immediately since it's gonna be used
+                lockBeforeWrite: true,
+                // Always write, even if it's already persisted for consistency.
+                // The in-memory validatorStore is the ground truth to decide duplicates
+                persistIfDuplicate: true,
+              });
+
+              // Add to in-memory store to start validating immediately
+              this.validator.validatorStore.addSigner({type: SignerType.Local, secretKey});
+
+              statuses[i] = {status: ImportStatus.imported};
+            } catch (e) {
+              statuses[i] = {status: ImportStatus.error, message: (e as Error).message};
+            }
+          },
+          (e: Error) => {
+            statuses[i] = {status: ImportStatus.error, message: e.message};
+          }
+        );
       } catch (e) {
         statuses[i] = {status: ImportStatus.error, message: (e as Error).message};
       }
```

Comment:
Why the two catches here.  Could we have this catch in `queue` by making it async and awaiting the tasks and passing errors sync/async to `onError`

https://github.com/ChainSafe/lodestar/blob/d89400cc46525061304a4eabdce1ea860d2f617f/packages/cli/src/cmds/validator/keymanager/decryptKeystores/threadPool.ts#L43-45

Reply:
I am not a big fan of having these 3 error handlers/catch blocks here but each of them handles different error cases

Lets just look at what each of those cases are

---
**The outer try-catch:** https://github.com/ChainSafe/lodestar/blob/d89400cc46525061304a4eabdce1ea860d2f617f/packages/cli/src/cmds/validator/keymanager/impl.ts#L178
1. no password error (L136)
2. parse error (L139)
3. getPubkeyHexFromKeystore error (L140)
4. and very unlikely queue errors (L148)

I think could move that block up to only handle cases 1-3 as it is unlikely that `queue` throws an error

---
**The `onError` handler:** https://github.com/ChainSafe/lodestar/blob/d89400cc46525061304a4eabdce1ea860d2f617f/packages/cli/src/cmds/validator/keymanager/impl.ts#L174

This one just catches any errors that are thrown by the worker that decrypt the keystore e.g. invalid password

---
**The inner try-catch:** https://github.com/ChainSafe/lodestar/blob/d89400cc46525061304a4eabdce1ea860d2f617f/packages/cli/src/cmds/validator/keymanager/impl.ts#L170

Catches error due to failure to persist keystore to disk `writeKeystore` (L156)

---

Is your idea to make `onDecrypted` callback async? If that is the case I don't think `onError` would catch it unless we would do something like this

```ts
task.then(result => {
  try {
    onDecrypted(result);
  } catch (error) {
    onError(error);
  }
}).catch(onError);
```

Maybe would help to add some comments to individual error handlers to make it bit easier to follow

Reply:
ok nvm my comment ☝️  I think you are right can just handle it in `onError` handler and don' even need to make it async (it is wrapped into a promise automagically)